text/0000-deref-move.md

```diff
@@ -0,0 +1,325 @@
+
+# Summary
+[summary]: #summary
+
+Add a new `DerefMove` trait that allows consuming a smart pointer to move its
+contents, as in `let x = *p;`.
+
+For more background and discussion, see #178 and #997.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, two smart pointer traits provide access to values contained in smart
+pointers: `Deref` and `DerefMut`. These traits allow overloading of the
+dereferencing (unary `*`) operator, and by extension participate in autoderef in
+a number of places, such as method calls and index expressions.
+
+These two traits, however, only dereference references to produce references to
+the contained value. They do not offer the option to dereference by value,
+consuming the smart pointer to produce the owned value inside. As a special
+case, `Box<T>` has compiler support allowing it to be dereferenced (explicitly
+or implicitly) to produce the inner `T`. Because this is special-cased in the
+compiler, this functionality is not available for other smart pointer types.
+
+Other smart pointers may support this consuming operation, but it must be done
+explicitly due to the lack of language support. For instance, in the stdlib, the
+following methods both consume a smart pointer to produce the inner value:
+
+* `ManuallyDrop<T>::into_inner()`
+* `binary_heap::PeakMut<'a, T>::pop()`
+* `Cow<'a, B>::into_owned()`
+
+I have not made any attempt to survey into the larger ecosystem for more
+examples, but there are probably many more.
+
+Note that removing `FnBox` is explicitly *not* part of the motivation for this
+RFC. It has been suggested in a few places, such as in discussion of
+rust-lang/rust#28796, that `DerefMove` would allow `Box<FnOnce()>` to be called
+without needing the `FnBox` workaround. This is not the case, however: the inner
+value is dynamically-sized and thus cannot be moved onto the stack. Indeed, one
+can call a `Box<T>` where `T` is a concrete type implementing `FnOnce()`.
+`DerefMove` would allow `Box<FnOnce()>` to continue not working, just without a
+special case in the compiler to allow it to do so.
+
+This will not remove all of `Box`'s special cases from the compiler, but
+represents progress towards that goal.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The following trait is defined in `std::ops`:
+
+```rust
+trait DerefMove : DerefMut {
+    fn deref_move(self) -> Self::Target;
+}
+```
+
+This trait allows moving out of the result of a derefence. This can happen
+either explicitly, as in `let x = *p;`, or implicitly, as in `p.into_foo()`.
+`Box<T>` from the standard library implements `DerefMove`, allowing you to move
+out of it.
+
+```rust
+let b = Box::new("s".to_owned());
+let s = *b;
+// ERROR: b has already been moved on the previous line.
+// let t = *b;
+```
+
+If the `Target` type is `Copy`, then `DerefMove` will not be used, and instead
+`Deref` will be used and the resulting referenced value will be copied:
```

Comment:
I think understand why this is necessary after staring at it for a while (since otherwise existing code that expects to be able to copy out of `*x` would be broken as it would move out of `b` incorrectly).  But as I noted above: it's not clear to me why the `Target` of `DerefMove` has to be the same as that for `Deref` since the rule seems to be that if the target for `Deref` is `Copy` we ignore `DerefMove`.  You could always lint for implementations of `DerefMove` with targets that are known to be `Copy` at definition time.